Fix: update banner issue fixes #138

subhayand10 opened the pull request:
The register service worker logic only runs once when we land on home page.
So, whenever there's a new update and land on home page, we see the update banner(toast) as the register logic has been re initialized.
When we navigate to other pages, the page does not run the register logic again in the same lifecycle context as the home page. As a result, the toast is not shown.

Fix:
Used localStorage (which acts as a global context for all the pages) to store a flag. When the flag is true, we show the banner and when we have clicked the toast and updated the app, the flag is removed.

nyakotey reviewed (approved):
@RDjarbeng

RDjarbeng commented:
Great, will review

RDjarbeng commented:
Code looks okay but since I can't push to this fork I can't test to see if the update banner actually appears

RDjarbeng commented:
Okay, cloned the fork to test out. The update banner shows on the other pages now but there are 2 banners on the homepage now. See screenshot below @subhayand10 

![image](https://github.com/user-attachments/assets/82eda535-d339-4e04-9169-05ff2b0a3a29)

RDjarbeng reviewed (changes requested):
Needs to fix the double update banner on the homepage

RDjarbeng commented:
I think this double banner occurs because showUpdateToast() is called twice. Somehow both are triggered on the home page
here

```
// Check for stored update status
function checkForUpdateToast() {
    if (localStorage.getItem("sw-update-available") === "true") {
        showUpdateToast();
    }

```
And here
```
onNeedRefresh() {
        console.log("New app version, now available");
        localStorage.setItem("sw-update-available", "true"); 
        showUpdateToast(); 
    },
```

subhayand10 commented:
> I think this double banner occurs because showUpdateToast() is called twice. Somehow both are triggered on the home page here
> 
> ```
> // Check for stored update status
> function checkForUpdateToast() {
>     if (localStorage.getItem("sw-update-available") === "true") {
>         showUpdateToast();
>     }
> ```
> 
> And here
> 
> ```
> onNeedRefresh() {
>         console.log("New app version, now available");
>         localStorage.setItem("sw-update-available", "true"); 
>         showUpdateToast(); 
>     },
> ```

You are right the double banner happens on the home page because of this. But when I was testing my code thoroughly after you found out this, I also found out some other issues which arises when we're using local Storage to show banners.

One of the issues being whenever we reject/ don't update our app and close the browser and next time when we open our application, the application comes as the updated one (which is normal as browsers gives us the updated version of websites/service worker versions when we open it).

Another issue is when we have already one update pending and we push another update, double banner is shown. Browsers always merges the latest service worker update, so every time we need to show only one banner, old updates automatically gets overridden in browser.

So, I've removed local storage and added a new logic where we are checking for the waiting service worker updates on page load and also when a new update comes in on top of the old, unmerged update. Also, we always limit the banner to one with a countBanner variable.

**Notes for testing: -**

Step 1: - In local, create a new build with new code and either update the app from banner (if it comes) or from devtools->service worker (service worker update).

Step 2: - Now you have the latest app, and you can start testing by making small changes to code and then rebuilding

RDjarbeng commented:
Alright will test the new changes

RDjarbeng commented:
Thanks @subhayand10
